src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+                               val matchReverseComplement: Boolean,
+                               val readStructures: ReadStructure*) extends Iterator[AssignedRecords] {
+    /** The barcode string for reads that do not match. */
+    private val noMatchBarcode: String = {
+      readStructures.flatMap {
+        readStructure => // get the sample barcode read structure for this read.
+          val rs = readStructure.sampleBarcodeStructure(resetOffsets=true)
+          if (rs.isEmpty) None
+          else Some(rs)
+        }
+        .map { readStructure =>
+          StringUtil.repeatCharNTimes('N', readStructure.map(_.length).sum)
+        }.mkString(SampleBarcodeDelimiter)
+    }
+
+    /** Read structures for the template bases, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val templateReadStructures = readStructures.map { readStructure =>
+      readStructure.templateStructure(resetOffsets=false)
+    }
+
+    /** Read structures for the sample barcodes, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val sampleBarcodeReadStructures = readStructures.map { readStructure =>
+     readStructure.sampleBarcodeStructure(resetOffsets=false)
+    }
+
+    /** Read structures for the molecular barcodes, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val molecularBarcodeReadStructures = readStructures.map { readStructure =>
+      readStructure.molecularBarcodeStructure(resetOffsets=false)
+    }
+
+    /** Sample barcode to its associate metric */
+    private val barcodeToMetrics: Map[String, SampleBarcodeMetric] = {
+      val metricMap = new mutable.LinkedHashMap[String, SampleBarcodeMetric]()
+      sampleSheet.foreach { sample =>
+        val barcode: String = sample.sampleBarcode match {
+          case Some(b) => b.concatenatedBarcode
+          case None    => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a sampleBarcode")
+        }
+        val libraryId: String = sample.libraryId match {
+          case Some(id) => id
+          case None     => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a library id")
+        }
+        metricMap.put(barcode, SampleBarcodeMetric(barcodeName=sample.sampleName, libraryName=libraryId, barcode=barcode))
+      }
+      metricMap.put(noMatchBarcode, SampleBarcodeMetric(UnmatchedSampleId, UnmatchedSampleId, noMatchBarcode))
+      metricMap.toMap
+    }
+
+    def hasNext: Boolean = input.hasNext
+
+    def next(): AssignedRecords = {
+      if (!hasNext) throw new NoSuchElementException("Calling next() when hasNext is false")
+
+      val fastqRecords                        = input.next()
+      val sampleBarcodeReadBases: Seq[String] = getSampleBarcode(fastqRecords).filter(_.nonEmpty)
+      val sampleOrdinal: Int                  = getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases.mkString, sampleSheet).getOrElse(headers.length)
```

Comment:
The `get` prefix here made me think this is a simple lookup, when this is actually where all the matching takes place.